pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,564 @@
+}
+
+func (plugin *storageosPlugin) RequiresRemount() bool {
+	return false
+}
+
+func (plugin *storageosPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
+	return []v1.PersistentVolumeAccessMode{
+		v1.ReadWriteOnce,
+		v1.ReadOnlyMany,
+	}
+}
+
+func (plugin *storageosPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
+	return plugin.newMounterInternal(spec, pod, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, manager storageosManager, mounter mount.Interface) (volume.Mounter, error) {
+	source, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageosMounter{
+		storageos: &storageos{
+			podUID:          pod.UID,
+			pvName:          spec.Name(),
+			volName:         source.VolumeName,
+			namespace:       source.Namespace,
+			pool:            source.Pool,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(pod.UID, source.Namespace, source.VolumeName, spec.Name(), plugin.host)),
+		},
+		fsType:      source.FSType,
+		readOnly:    readOnly,
+		devicePath:  storageosDevicePath,
+		diskMounter: &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewUnmounter(pvName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(pvName, podUID, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newUnmounterInternal(pvName string, podUID types.UID, manager storageosManager, mounter mount.Interface) (volume.Unmounter, error) {
+
+	// Parse volume namespace & name from mountpoint if mounted
+	namespace, volName, err := getVolumeInfo(pvName, podUID, plugin.host)
+	if err != nil {
+		return &storageosUnmounter{storageos: &storageos{}}, nil
+	}
+
+	return &storageosUnmounter{
+		storageos: &storageos{
+			podUID:          podUID,
+			pvName:          pvName,
+			volName:         volName,
+			namespace:       namespace,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, namespace, volName, pvName, plugin.host)),
+		},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec, &storageosUtil{})
+}
+
+func (plugin *storageosPlugin) newDeleterInternal(spec *volume.Spec, manager storageosManager) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.StorageOS == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.StorageOS is nil")
+	}
+	return &storageosDeleter{
+		storageos: &storageos{
+			pvName:    spec.Name(),
+			volName:   spec.PersistentVolume.Spec.StorageOS.VolumeName,
+			namespace: spec.PersistentVolume.Spec.StorageOS.Namespace,
```

Comment:
This is the volume's namespace, not necessarily the Pod's.  Starting to think I should rename it volNamespace to make it more clear - what do you think?